sycl/plugins/esimd_cpu/esimd_libcm.hpp

```diff
@@ -0,0 +1,327 @@
+
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::item<1, true>>::value>::type
+  run() {
+    auto WrappedLambda_ITEM_OFFSET_1DIM =
+        makeWrapper_ITEM_OFFSET_1DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ITEM_OFFSET_1DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ITEM_OFFSET_1DIM),
+                      WrappedLambda_ITEM_OFFSET_1DIM.get());
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::item<2, true>>::value>::type
+  run() {
+    auto WrappedLambda_ITEM_OFFSET_2DIM =
+        makeWrapper_ITEM_OFFSET_2DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ITEM_OFFSET_2DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ITEM_OFFSET_2DIM),
+                      WrappedLambda_ITEM_OFFSET_2DIM.get());
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::item<3, true>>::value>::type
+  run() {
+    auto WrappedLambda_ITEM_OFFSET_3DIM =
+        makeWrapper_ITEM_OFFSET_3DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ITEM_OFFSET_3DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ITEM_OFFSET_3DIM),
+                      WrappedLambda_ITEM_OFFSET_3DIM.get());
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::nd_item<1>>::value>::type
+  run() {
+    auto WrappedLambda_NDITEM_1DIM =
+        makeWrapper_NDITEM_1DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_NDITEM_1DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_1DIM),
+                      WrappedLambda_NDITEM_1DIM.get());
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::nd_item<2>>::value>::type
+  run() {
+    auto WrappedLambda_NDITEM_2DIM =
+        makeWrapper_NDITEM_2DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_NDITEM_2DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_2DIM),
+                      WrappedLambda_NDITEM_2DIM.get());
+    ArgVector.clear();
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::nd_item<3>>::value>::type
+  run() {
+    auto WrappedLambda_NDITEM_3DIM =
+        makeWrapper_NDITEM_3DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_NDITEM_3DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_NDITEM_3DIM),
+                      WrappedLambda_NDITEM_3DIM.get());
+    ArgVector.clear();
+  }
+};
+
+inline void __cm_mt_barrier() { cm_support::mt_barrier(); }
+
+inline void __cm_set_slm_size(size_t size) { cm_support::set_slm_size(size); }
```

Comment:
It is not clear why `__cm` prefix is used here. What is the purpose of these functions?

Reply:
Like mentioned above, they are called from ESIMD intrinsic implementations.

Reply:
Looks like we noticed an elephant in the room :) I envision a couple of problems with this:
- `__cm_set_slm_size` is a device-side API unlike kernel launching and they should not be intermingled within the same header
- This will require linkage of user application to the libCM where `cm_support::set_slm_size` is implemented, even if emulation is not needed.
@romanovvlad, @smaslov-intel, we need to decide what is acceptable here. We obviously don't want to link with libCM always. So I see 2 variants (maybe you can see more)
- require the user to compile and link differently when EMU support is wanted:
    -    add -DSUPPORT_CPU_EMU to compilation under which all `__esimd*` intrinsics will call libcm APIs when EMU is enabled
    -    add -lcm to link command
- introduce a CM device interface table dynamically initialized by the SYCL RT. The interface table will be obtained from the EMU plugin. We can discuss required PI extensions if needed.
Red arrow below shows the second approach.

![image](https://user-images.githubusercontent.com/23384882/106980436-07564300-6715-11eb-86f9-ef7d160e2c6f.png)

Reply:
How does a kernel ends up having direct calls to `__cm_set_slm_size` or `__esimd_intrinsicX`?

Reply:
@romanovvlad In my implementation prepared for later PR-posting, following functions are calling `__cm_*` functions through a couple of wrapping functions. The 'caller' functions are defined in '/sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp'. `__cm_*` functions are mostly called in 'esimd_memory_intrin.hpp'.

```
slm_init(size)
 ->__cm_set_slm_size(size)
esimd_barrier()
 -> __cm_mt_barrier()
slm_load(offsets, predicates)
 -> __cm_get_slm()
slm_load4(offsets, predicates)
 -> __cm_get_slm()
slm_block_load(offset)
 -> __cm_get_slm()
slm_store(values, offsets, predicates)
 -> __cm_get_slm()
slm_store4(values, offsets, predicates)
 -> __cm_get_slm()
slm_block_store(offset, values)
 -> __cm_get_slm()
```

Reply:
right, the call is not direct of course, but through wrappers, which I omitted for clarity. The picture only shows the essence.

Upon further thinking, my preference would be some solution which does not require recompilation, as this is not always possible. So from the two variants I suggested above I prefer the second or any other w/o recompilation.

Reply:
Hm... It would be nice if calls to those functions are done thru a kernel argument like nd_item which could have virtual table. But this should break ABI.
As for "introduce a CM device interface table dynamically initialized by the SYCL RT." approach, since it requires just late binding rather than dynamic dispatching to separate implementations, can we just keep those symbols unresolved during compilation so they are automatically resolved during call to dlopen(..., RTLD_NOW | RTLD_GLOBAL) ? Not sure we can do the same on windows though.

Reply:
Calling through a kernel argument would be ideal, but is not feasible. There are only few APIs which map to the standard SYCL APIs. Many other (like accessing memory through accessors) don't map, so custom (ESIMD-specific) device-side API is used.

Regarding keeping symbols unresolved with automatic resolution upon `dlopen` - I'd be fine with that if it worked reliably on all platforms. I actually wasn't aware that this was possible even on Linux. @sndmitriev - do you know whether this approach works reliably on Linux and Windows?

Reply:
Here is my suggestion in more details:
```cpp
struct ESIMDDeviceInterface {
  virtual void barrier() = 0;
  virtual void set_slm_size(size_t size) = 0;
};

// PI interface (name TBD):
void* piextGetPluginSpecificData();

// Implementation in the plugin:
struct ESIMDDeviceInterfaceImpl : public ESIMDDeviceInterface {
  void barrier() override {
    cm_support::mt_barrier();
  }
  void set_slm_size(size_t size) override {
    cm_support::set_slm_size(size);
  }
};

void* piextGetPluginSpecificData() {
  static ESIMDDeviceInterfaceImpl* ESIMDIntf = nullptr;

  if (!ESIMDIntf)
    ESIMDIntf = new ESIMDDeviceInterfaceImpl();
  return ESIMDIntf;
}

// SYCL Rt changes:
std::map<backend, void*> PluginOpaqueDataRegistry;

void* getPluginSpecificData(backend be) { return PluginOpaqueDataRegistry[be]; }

// somewhere in SYCL RT:
{
  PluginOpaqueDataRegistry[esimd_emu] = esimd_plugin.piextGetPluginSpecificData();
}

// intrinsic implementaiton:
#ifndef __SYCL_DEVICE_ONLY__
void __esimd_barrier() {
  ESIMDDeviceInterface *I = reinterpret_cast<ESIMDDeviceInterface*>(getPluginSpecificData(esimd_emu));
  I->barrier();
}
#endif
```

Reply:
We decided that we are going to follow what @romanovvlad suggested - delayed-library loading.

In Linux, this delayed-library loading is supported with '```dlopen([LIB_NAME],[LOADING_OPTION])```' calls from c/c++ files along with '```-ldl```' command line option as POSIX standard.

In Windows, delayed loading is supported with compiler command line argument('```/DELAYLOAD:[LIB_NAME]```') while it seems that there is no c/c++ API equivalent to '```dlopen()```'.

(There is an active open-source project enabling 'dl*' functionalities under Win32 environment, which could imply that Microsoft does not support the functionalities officially yet. (https://github.com/dlfcn-win32/dlfcn-win32))

Reply:
I tested delayed-library loading on Windows environment.

Without delayed-library loading, the command line for compilation looks like follows
```%LLVM_ROOT%\bin\clang++ -fsycl -fsycl-explicit-simd -llibcm -I %LLVM_ROOT%\include\sycl -g -std=c++17 -o .\vadd_1d.exe .\vadd_1d.cpp```

In order to enable delayed-library loading, following two lines have to be added in application file (such as vadd_1d.cpp). This is an alternative approach to specifying library(ies) to be delay-loaded instead of ```[LIB_NAME] delayimp.lib``` in command line.
```#pragma comment(lib, "delayimp")```
```#pragma comment(lib, "libcm")```

With these two lines in application source file, following command line works.
```%LLVM_ROOT%\bin\clang++ -fsycl -fsycl-explicit-simd -I %LLVM_ROOT%\include\sycl -g -std=c++17 -o .\vadd_1d.exe .\vadd_1d.cpp```

I don't know why it is working without ```link /delayload:libcm.dll``` command line argument.

Reply:
> (There is an active open-source project enabling 'dl*' functionalities under Win32 environment, which could imply that Microsoft does not support the functionalities officially yet. (https://github.com/dlfcn-win32/dlfcn-win32))

I was wrong about this statement - there is C/C++ API equivalent to 'dlopen' in Windows - LoadLibrary. Its usage is almost  same as 'dlopen' - (1) open library file / (2) search symbol from the file handle and get function pointer / (3) call function through function pointer.

Reply:
For the record - @dongkyunahn-intel tried the delayed loading approach on Windows - did not work w/o explicit GetProcaddress-based binding. We decided to go with the interface table.